File: 78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python
#
# [78] 子集
#
# https://leetcode-cn.com/problems/subsets/description/
#
# algorithms
# Medium (77.16%)
# Likes:    593
# Dislikes: 0
# Total Accepted:    95.2K
# Total Submissions: 123K
# Testcase Example:  '[1,2,3]'
#
# 给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
#
# 说明：解集不能包含重复的子集。
#
# 示例:
#
# 输入: nums = [1,2,3]
# 输出:
# [
# ⁠ [3],
# [1],
# [2],
# [1,2,3],
# [1,3],
# [2,3],
# [1,2],
# []
# ]
#
#

# @lc code=start


class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # 回溯
        ans = []
        l = len(nums)
        def search(target,ind):
            ans.append(target)
            for i in range(ind,l):
                search(target+[nums[i]],i+1)
        search([],0)
        return ans

        # 逐个元素添加
        ans = [[]]
        for i in nums:
            ans += [x+[i] for x in ans]
        return ans

        # 二进制掩码
        ans = []
        l = len(nums)
        max_bit = 1 << l
        for i in range(max_bit):
            mask = bin(i | max_bit)[3:]
            ans.append([nums[x] for x in range(l) if mask[x] == '1'])
        return ans


# @lc code=end

File: 78.子集.go
/*
 * @lc app=leetcode.cn id=78 lang=golang
 *
 * [78] 子集
 *
 * https://leetcode.cn/problems/subsets/description/
 *
 * algorithms
 * Medium (81.11%)
 * Likes:    2090
 * Dislikes: 0
 * Total Accepted:    648K
 * Total Submissions: 798.8K
 * Testcase Example:  '[1,2,3]'
 *
 * 给你一个整数数组 nums ，数组中的元素 互不相同 。返回该数组所有可能的子集（幂集）。
 *
 * 解集 不能 包含重复的子集。你可以按 任意顺序 返回解集。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,3]
 * 输出：[[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [0]
 * 输出：[[],[0]]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * -10
 * nums 中的所有元素 互不相同
 *
 *
 */
package main

// @lc code=start
func subsets(nums []int) [][]int {
	ans := [][]int{}
	var backTrack func(int, []int)
	backTrack = func(ind int, val []int) {
		ans = append(ans, val)
		for i := ind; i < len(nums); i++ {
			backTrack(i+1, append([]int{nums[i]},val...))
		}
	}
	backTrack(0, []int{})
	return ans
}

func subsets1(nums []int) [][]int {
	ans := [][]int{[]int{}}
	for _, i := range nums {
		for _, a := range ans[:] {
			ans = append(ans, append(append([]int{}, a...), i))
		}
	}
	return ans
}

// @lc code=end
